beetsplug/spotify.py

```diff
@@ -536,3 +556,71 @@ def _output_match_results(self, results):
             self._log.warning(
                 f'No {self.data_source} tracks found from beets query'
             )
+
+    def _fetch_info(self, items, write, force):
+        """Obtain track information from Spotify."""
+        spotify_audio_features = {
+            'acousticness': ['spotify_track_acousticness'],
+            'danceability': ['spotify_track_danceability'],
+            'energy': ['spotify_track_energy'],
+            'instrumentalness': ['spotify_track_instrumentalness'],
+            'key': ['spotify_track_key'],
+            'liveness': ['spotify_track_liveness'],
+            'loudness': ['spotify_track_loudness'],
+            'mode': ['spotify_track_mode'],
+            'speechiness': ['spotify_track_speechiness'],
+            'tempo': ['spotify_track_tempo'],
+            'time_signature': ['spotify_track_time_sig'],
+            'valence': ['spotify_track_valence'],
+        }
+        import time
+
+        no_items = len(items)
+        self._log.info('Total {} tracks', no_items)
+
+        for index, item in enumerate(items, start=1):
+            time.sleep(.5)
+            self._log.info('Processing {}/{} tracks - {} ',
+                           index, no_items, item)
+            try:
+                # If we're not forcing re-downloading for all tracks, check
+                # whether the popularity data is already present
+                if not force:
+                    spotify_track_popularity = \
+                        item.get('spotify_track_popularity', '')
+                    if spotify_track_popularity:
+                        self._log.debug('Popularity already present for: {}',
+                                        item)
+                        continue
+
+                popularity = self.track_popularity(item.spotify_track_id)
+                item['spotify_track_popularity'] = popularity
+                audio_features = \
+                    self.track_audio_features(item.spotify_track_id)
+                for feature in audio_features.keys():
+                    if feature in spotify_audio_features.keys():
+                        item[spotify_audio_features[feature][0]] = \
+                            audio_features[feature]
+                item.store()
+                if write:
+                    item.try_write()
+            except AttributeError:
+                self._log.debug('No track_id present for: {}', item)
+                pass
```

Comment:
I think his `try`/`except` handler should probably go _just_ around the `item.spotify_track_id` access above, to make it clear to readers that this is the only thing we expect to fail.

Reply:
This doesn't seem to be addressed in your revised PR, was that intentional? For which reason?

Reply:
@wisp3rwind I am not sure exactly what is expected here. I could move [this block](https://github.com/beetbox/beets/blob/7e648db2583774e80834642b3a75ded7b3fa25ab/beetsplug/spotify.py#L588-L594)  outside the `try` block. Would that help?

Reply:
I think what @sampsyo meant is this:

```python
            try:
                spotify_track_id = item.spotify_track_id
            except AttributeError:
                self._log.debug('No track_id present for: {}', item)
                continue
            
            popularity = self.track_popularity(spotify_track_id)
            item['spotify_track_popularity'] = popularity
            audio_features = \
                self.track_audio_features(spotify_track_id)
            for feature in audio_features.keys():
                if feature in self.spotify_audio_features.keys():
                    item[self.spotify_audio_features[feature]] = \
                        audio_features[feature]
            item.store()
            if write:
                item.try_write()
```